### ISSUE TITLE
Two New AIX Machine Issues (cmake)

### ISSUE THREAD

**Issue by rajdeepsingh1** [eclipse/omr#5016]
FYI @rwy0717 

Receiving the following compile issue on two new AIX machines. Looks to be a cmake issue however we cannot downgrade due to dependencies. 

`omr-aix71p8-1`
`omr-aix71p8-2 `

```
Scanning dependencies of target gcexample
[ 33%] Building CXX object example/CMakeFiles/gcexample.dir/gcmain.cpp.o
[ 33%] Building CXX object example/CMakeFiles/gcexample.dir/__/util/main_function/main_function.cpp.o
[ 33%] Linking CXX executable gcexample
[ 33%] Built target gcexample
[ 33%] Building CXX object omrsigcompat/CMakeFiles/omrsig.dir/omrsig.cpp.o
[ 33%] Linking CXX shared library libomrsig.so
Error running link command: No such file or directory
make[2]: *** [omrsigcompat/CMakeFiles/omrsig.dir/build.make:90: omrsigcompat/libomrsig.so] Error 2
make[1]: *** [CMakeFiles/Makefile2:1656: omrsigcompat/CMakeFiles/omrsig.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

**Comment by rwy0717** [comment]
Can you post the full `cmake` and `make VERBOSE=1` log? Maybe you can put it in a [gist](https://gist.github.com/).

**Comment by jdekonin** [comment]
cmake / setup
https://gist.github.com/jdekonin/4228bc0d3aa769a319c8a2f607a15045
make VERBOSE=1
https://gist.github.com/jdekonin/cbabb3d8c322c711d570857021edcf40

Note : 
- cmake is v3.16, not v3.14 as on the 2 working systems.  yum isn't permitting a downgrade to that version though
- xlc is v13.1.3.5, not v13.1.3.0 as on the 3 working systems.  But we fail in the same manner on an internal system I use (aix71p7rt2) with the v13.1.3.0 and cmake v3.16.

**Comment by jdekonin** [comment]
@rajdeepsingh1 had you tested either new OMR machine with an OpenJDK8-OpenJ9 compile?

**Comment by jdekonin** [comment]
I just tested a OpenJDK8-OpenJ9 and OpenJDK11-OpenJ9 builds on omr-aix71p8-2, OpenJDK14-OpenJ9 is running now, but its already compiled the OMR portion of the compile.  I think the difference is that OpenJ9 triggers configure and not cmake.  omr-aix71p8-1 and omr-aix71p8-2 should be identical.

@rwy0717 is there a OMR support statement for which version of cmake AIX should have installed?  I'd like to go with whatever we can get through the official channels which with the AIX toolkit is v3.16.

I can see this was added March 24th and this version made cmake-filesystem and cmake-rpm-macros packages obsolete.  Not sure if this is a defect in the new cmake or if this is a problem on the OMR source side.  Please advise.

**Comment by rwy0717** [comment]
>  I think the difference is that OpenJ9 triggers configure and not cmake. 

Yep, openj9 AIX builds are still using the configure-based build system.

> @rwy0717 is there a OMR support statement for which version of cmake AIX should have installed? I'd like to go with whatever we can get through the official channels which with the AIX toolkit is v3.16.

The oldest version we support is 3.2. I would like to support every version newer than that, including 3.16 on AIX. CMake 3.16 changed **a lot** in how it compiles XLC, so I'm not surprised there is an issue. I believe this can be fixed in OMR.

**Comment by jdekonin** [comment]
To summarize our slack chat on the topic as we found a couple problems.

1. the path to CreateExportList is wrong. This file is usually included in the XLC installation and cmake is told to look here `/usr/vac/bin/CreateExportList` which is actually a xlc12 location.  For xlc13 its located here `/opt/IBM/xlC/13.1.3/bin/CreateExportList`.

2. As of 3.16, cmake no longer uses the CreateExportList script, it uses a built in script.  In the snippet of code for v3.16 installed on the machine it appears there are two lines that are not in the [original source](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/Platform/AIX-XL.cmake#L31-33).
```
    "<CMAKE_AR> -X32_64 qc <TARGET> <TARGET_SONAME>"
    "strip  -e -X32_64 <TARGET_SONAME>"
```
 
Compiling my own version of cmake v3.16.5 didn't work so well as it encounters a seg fault.  I can look at revisiting this in the future, but my current priorities have that out a week or two at least.
```
-- The C compiler identification is XLClang 16.1.0.0
-- The CXX compiler identification is XLClang 16.1.0.0
-- Check for working C compiler: /opt/IBM/xlC/16.1.0/bin/xlclang
./bootstrap[1676]: 30474340 Segmentation fault(coredump)
---------------------------------------------
Error when bootstrapping CMake:
Problem while running initial CMake
```
So what I have done to resolve, is to open #5030 to reference the correct location of `CreateExportList` and added a previously used version of cmake v3.7.2 as downgrading via yum package manager was not possible.

**Comment by jdekonin** [comment]
I’ve marked omr_ci_1 and omr_ci_2 offline in jenkins.

I have added omr-aix71p8-1 and omr-aix71p8-2 as replacements.  Both have been tested with last nights build (successfully of course). 

Let me know if there are any problems with either omr-aix71p8-1 or omr-aix71p8-2. I plan on removing omr_ci_1 and omr_ci_2 from jenkins Thursday 5pm unless there are issue with these new additions.

I'll also try follow up with the AIX Toolkit for Linux folks wrt the v3.16 code problem.

**Comment by dnakamura** [comment]
I did a bit of digging an it looks like the segfault isssue is a result of our old friend, references outliving their referenced object (#4540), although in this case I need to do a bit of digging to see if it cmakes code is allowed by the c++ spec. Building with optimizations enabled gets rid of the segfault, but then causes an error because `<CXX_LINK_FLAGS>` doesnt get properly replaced in the link command, which also smells of an issue with string referces persisting past the life of the string. Also it seems that xlcang doesn't support the `-qlanglvl=tempsaslocals` which was very helpful in diagnosing these issues.  After a bit of digging I noticed this in a readme file in the source RPM for cmake

> Xlclang can probably compile CMake and can be used with, but we don’t use it, so we will ignore this possibility.

**Comment by jdekonin** [comment]
Devin supplied a compiled v3.17.1 that I have tested on omr_ci_1.  I will put into place to make this the new default cmake unless there is a reason not too.  @rwy0717 @dnakamura are you both fine with this?

**Comment by jdekonin** [comment]
Discussed with @rwy0717 via slack and received go ahead.  Both machines have v3.17.1 now on the PATH.

I believe this issue can be closed.  #5116 should be able to proceed.
